Pa accordian

aschwanden1 opened the pull request.

hategan reviewed (commented):
Thank you Pascal for all the effort you have put into this.

I do have some inline comments. Some are carried over from #24. My stance has not changed on those. Others are easy to fix and some are probably not as easy. One straightforward thing is that various OS special files and, in most cases, binary files (e.g., `.DS_Store`) should not be in git.

I also mentioned in #24 why big PRs are problematic as well as the fact that we should discuss big changes with the team and community so that we increase the chances that the team and community at large can converge on a generally agreeable solution. That remains relevant.

aschwanden1 commented on src/psij/web/js/SitesModel.js:
```diff
@@ -0,0 +1,381 @@
+
+        getMonthAndDay_( PS.sites );
+
+        var CS = makeModelInstance_( CONF.curated_sites );
+        var CS2 = makeModelInstance_( CONF.curated_sites2 );
```
One is for PSIJ and the other is for SDK.

hategan replied:
There can be different sets based on customization.js, but not sure about two concurrently in the same instance. In other words, I don't see why there should be one for the SDK in the PSI/J dashboard or one for PSI/J in the SDK dashboard.

aschwanden1 commented on src/psij/web/js/IntroBlurb.js:
```diff
@@ -0,0 +1,38 @@
+var VC = VC || {};
+
+VC.IntroBlurb = Vue.component("intro-blurb", {
+    data: function () {
```
I was told that they wanted the blurb to dissappear and not reappear after people closed them, hence the need for persistance.

hategan replied:
I am not sure who "they" is. We're a distributed project with multiple team members in multiple organizations. And we want this to be an open community project. The concept of some fuzzy "they" is incompatible with that. I don't think this was necessarily communicated clearly, and I apologize for that.

Nonetheless, I think that whatever it is, it still needs to make sense. And this is a rather unique idea that I haven't quite seen except (and I might have mentioned this in #24) in nag screens and when mandated by some law, as in the cookie preferences thing that has been popping up recently. However, a headline of what the current page is is not that.

So I'd be fine with this, but it's not done in other places and it does not improve in a clear way, all while adding code. It does not hence seem justified.

aschwanden1 replied:
If i remember correctly, I think Dan asked for this feature.

hategan replied:
Well, we probably need to decide, as a community and a team, whether that should override the need for a discussion and agreement. What we last settled on and codified was https://github.com/ExaWorks/psij-python/blob/main/CONTRIBUTING.md.

aschwanden1 commented on src/psij/web/js/RowAccordian.js:
```diff
@@ -0,0 +1,66 @@
+};
+
+
+VC.RowAccordian = Vue.component("row-accordian", {
```
the accordian we need, needs to have rows and columns that span across the body of the accordian to the clickable row of the accordian.  Also, we need the arrow of the accordian to change direction when you click on it.

hategan replied:
Sorry, but I disagree, just as I did before. There are expansion panels (https://vuetifyjs.com/en/components/expansion-panels/#usage) in vuetify which can be used instead. The icons can be customized, so that is not an issue.

The content of the expanders need not be table rows, but I do not see why they couldn't be.

However, even if so, vuetify has tables where rows can be grouped or expanded: 
https://vuetifyjs.com/en/components/data-tables/#grouping 
https://vuetifyjs.com/en/components/data-tables/#expandable-rows

hategan replied:
I will resolve this for now. For reference, I do not think the reasoning behind having a custom component applies much since the current solution (with the custom components) is not to expand rows in a table, but to have multiple tables that happen to align perfectly thus giving the appearance of a single table. But this takes extra effort to ensure that the columns form multiple tables are aligned and could as well be done better with plain divs.

aschwanden1 commented on src/psij/web/css/whole.css:
```diff
@@ -0,0 +1,2 @@
+@import url("https://fonts.googleapis.com/css?family=Source+Sans+Pro:300,300italic,400,400italic,600");.RowAccordian{border-top:1px solid #e0e0e1}.RowAccordian td{width:90px;padding:5px;position:relative}.RowAccordian td:first-child{width:345px;cursor:pointer;color:#333333}.RowAccordian .logo_name{font-weight:bold}.RowAccordian .ArrowDown{width:0;height:0;display:inline-block;border-left:8px soli ...
```
I agree with the simplicity principle.  But, SCSS asks so little of us and gives back so much more.  Whereas something like Vue or other "modern" javascript frameworks are incredibly complex and give back very little advantage.  If anything, I think we should use less Vue.

hategan replied:
I think I explained rather clearly and with specifics why SCSS is a problem. Adding npm to the list of dependencies is not "little". And you are not explaining how SCSS is "much more". Does it result in a significant simplification of our css files?

Adding Vue to the conversation or challenging the usefulness of widely used libraries or of all modern JS frameworks is not helpful.

aschwanden1 replied:
The reason I brought up Vue was because I see Vue as being far more complex than SCSS.  SCSS is very light weight and helps simplify things.


Here are a few of the benefits of SCSS:
* You can nest selectors and so you can specify all the CSS selectors underneath another one and easily see which rules are under which divs
* You can specify site wide color variables 
* We get aggregation and minification (something we want anyway) by default so our site loads faster (fewer http requests and smaller files).  and don't worry the browser is able to unpack it so it's still just as easy to debug

hategan replied:
Vue does not require me to install anything, whereas SCSS needs a compiler. You can probably use sass.js for that, which would be an improvement. However, it is still a technological change that should have some **specific** benefits **for this project**.

I am not looking for the "benefits of SCSS". I am looking, specifically, to see how adding this technology helps us. And what you list are not specific issues to the dashboard but general talking points for SCSS. Because you can also specify global color variables in plain CSS. As for loading time, all resources are loaded in parallel. Style.css takes about 66ms to load, whereas jquery takes 270 ms. You will not see one yota of benefit in making the CSS smaller unless it takes more than what the longest resource takes (so at least more than 270 ms).

aschwanden1 commented on src/psij/web/js/SitesModel.js:
```diff
@@ -0,0 +1,381 @@
+    };
+
+
+    var makeGoodStatesArray_ = function( entryCount ) {
+
+        var goods = [];
+        for( var x = 0; x < entryCount; x++ ) {
+            goods.push( PS.STATES.GOOD );
+        }
+
+        return goods;
+    };
```
that code doesn't work.  it just makes [undefined, undefined ....]

hategan replied:
I'm not sure about that: https://codepen.io/hategan/pen/XWBqjaW

aschwanden1 commented on src/psij/web/js/customization.js:
```diff
@@ -0,0 +1,125 @@
+                        scheduler: "SLURM"
+                    },
+                    "illyad.hidden.uoregon.edu": {
+                        scheduler: "Cron"
```
Yeah, those were just place holders.  I'll change it to SLURM.

aschwanden1 replied:
We'll probably need to verify what the actual schedulers are.

hategan replied:
The tests have that information. If you look, for example, here https://psij.testing.exaworks.org/run.html?site_id=cori08.nersc.gov&run_id=2023012509-a236657f00bc0379e9b5aa1619b2ede3596cafb0c42083438af3f2d4893c780d#branch-main, you will see that in `_discover_environment`, there is a `has_slurm` field that is set to `true`

hategan replied:
This are done manually, but should be fixed in a future PR. @aschwanden1, please open an issue to keep track of this.

hategan replied:
See #41

aschwanden1 commented on src/psij/web/js/customization.js:
```diff
@@ -0,0 +1,125 @@
+                    },
+                    "doesnotexistTEST.edu": {
+                        scheduler: "Cron"
+                    }
```
done

hategan commented on src/psij/web/js/customization.js:
```diff
@@ -29,22 +29,22 @@ var CONF = {
                     },
                     "illyad.hidden.uoregon.edu": {
-                        scheduler: "Cron"
+                        scheduler: "SLURM"
```
Those machines don't have a job scheduler. I think only axis at uoregon has a scheduler and it's slurm

aschwanden1 replied:
https://github.com/ExaWorks/psij-testing-service/issues/41

aschwanden1 commented on src/psij/web/js/BuildAccordianModelFromSites.js:
```diff
@@ -0,0 +1,96 @@
+    This builds the DataModel for the Calendar View.
+    It preps the Data for consumption.
+ */
+PS.BuildAccordianModelFromSites = function() {
```
PS is short for PSIJ   

I wanted to clean up the global namespace, so having a short 2 letter constant as the base for all our modules helps keeping the global namespace clean by making it easy to add modules underneath it.

hategan replied:
Keeping things under a single namespace is fine, and that should probably apply to everything.
"PSIJ" itself is short enough and less confusing.

hategan replied:
Not when it's confusing!

aschwanden1 commented on src/psij/web/js/lib.js:
```diff
@@ -385,7 +488,10 @@ var globalMethods = {
     breadcrumbs: function() {
-        var b = [{text: CUSTOMIZATION.projectName + " Tests", href: "summary.html"}];
+
+        var projName = PS.isSDKPage() ? "SDK" : "PSI/J";
```
I thought the psij and sdk page both use the same codebase.  are you suggesting that we should be loading sdk and psij from two different code deployments?

hategan replied:
customization.js, as the name suggests, is not part of the common code base. It is where we currently differentiate between PSI/J and SDK.

I think we discussed this a bit in #24 and, at that time, you convinced me that it would be a good idea to store project specific strings in a single place (like customization.js). At that time I was saying that, for the purpose of the introductory text, we could have two different html pages or some kind of templating for the text such that text changes are close to the view that represents them. I'm still somewhat in-between in that the text is part of the view, but I'm also leaning towards what you were saying about having all differences in one place.

However, code that explicitly mentions the SDK in a function name is a bit of a different thing and I do not think it is either what you were proposing then or what I was proposing then.

Specifically, `CUSTOMIZATION.projectName` is a more reasonable solution than `PS.isSDKPage() ? "SDK" : "PSI/J"`.

aschwanden1 replied:
okay but then we will need to have one customization.js for psij and one for SDK.   so do you just want me to make like: psij_customization.js and sdk_customization.js or did you have something else in mind?

hategan replied:
Let's do that. Have `instance/customization-psij.js` and `instance/customization-sdk.js`. Then, on the deployments, we do a symbolic link from either to `instance/customization.js` and we only load `instance/customization.js` in html files.

aschwanden1 replied:
I'm a bit worried that this won't work the same in dev as in prod hence opening us up to possibility of bugs, plus the additional step of setting up the symbolic link.

How about we do this:
<script>
var custSuffix = window.location.href.subdomain;
var custToLoad = "customization-" + custSuffix + ".js";
//  Now load the appropriate file:
</script>

hategan replied:
> I'm a bit worried that this won't work the same in dev as in prod

Why would that be the case?

> the additional step of setting up the symbolic link

It's a one time step.

> <script> var custSuffix = window.location.href.subdomain; var custToLoad = "customization-" + custSuffix + ".js"; // Now load the appropriate file: </script>

We don't know for sure that we'll keep the current url scheme.

aschwanden1 replied:
by what mechanism will the Link be different from psij to sdk?

hategan replied:
> by what mechanism will the Link be different from psij to sdk?

by running different `ln -s` commands when first setting up a deployment.

aschwanden1 replied:
so, it sounds like we'll need to create two locations, one for psij and one for sdk and two ln -s     and each subdomain will need to point to each codebase.  and symbolic links can break.  just seems more complicated and error prone.

hategan replied:
> it sounds like we'll need to create two locations, one for psij and one for sdk [...] and each subdomain will need to point to each codebase.

That's already the case. We have two containers each with an instance of the web service and database.

> symbolic links can break

I have not heard of this before. How do symbolic links break?

hategan commented on src/psij/web/js/common.js:
```diff
@@ -0,0 +1,19 @@
+
+PS.useLocalStaticDB = false;
+
+PS.getURL = function( po ) {
+
+    var firstThree = window.location.host.substr(0,3);
+    var strSDK = firstThree === "sdk" ? "sdk." : "";
+
+    return PS.useLocalStaticDB ? po : 'https://' + strSDK + 'testing.exaworks.org/' + po;
+};
```
Ca we maybe get rid of the sdk reference?

hategan commented on src/psij/web/js/lib.js:
```diff
@@ -60,22 +74,44 @@ var settingsDefaults = function(dict) {
-            }
-            else {
-                return val;
+
+            var sbb = "none found";
+
+            if( name === 'showBranchBubbles' ) {
+                sbb = localStorage.getItem('showBranchBubbles') === 'true' ? 'true' : false;
+            } else  if( name === 'inactiveTimeout' ) {
+                sbb = localStorage.getItem('inactiveTimeout') || 10;
             }
+
+            return sbb;
```
^^^

hategan replied:
Was there a need to specify default values here on top of initializing default values when the page was loaded?
What makes this code more robust, exactly?

aschwanden1 replied:
because that way you ensure that the value you get out is exactly one of those default values and nothing else.  the need is robustness.

hategan replied:
You get the default value in this function unless something is already stored in local storage.
So, if the first thing you do when you load lib.js is to store default values in local storage, will you ever get the default values from this function?

hategan replied:
Anyway, I don't want to drag this one, but no, the code is not more robust and this change should not have been here, since the settings are all initialize at page load. In fact, this is less robust because relies on hardcoded strings, fetches values twice for no good reason, and, for showBranchBubbles returns either the string 'true' or the boolean false, making it difficult to use in either a string or boolean expression.

hategan commented on sym-link.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+#  This is for setting up your sym links for dev.  It is needed for testing.
+#  You may modify the DIR below for your environment.
+#  enter 1 as first argument for PSIJ 
+PSIJ=$1
+
+DIR=~/psi-j-testing-service/src/psij/web/instance
+
+cd $DIR
+rm customization.js 
+
+if [ $PSIJ -eq 1 ]
+then
+   ln -s customization-psij.js customization.js
+else
+   ln -s customization-sdk.js customization.js
+fi
+
+ls -al $DIR
```
I don't think we need a script because this should not be something that needs to be done repeatedly. We should put the information that you could link one of the two on deployment in one of the relevant readme files.

hategan commented on src/psij/web/js/SitesModel.js:
```diff
@@ -0,0 +1,381 @@
+                highestDay_ = highestDay;
+            }
+        }
+    };
```
This was partially addressed. However, it's my understanding of the code that the curated sites page will show green for sites that have reported success at any time in the past.

hategan replied:
This is now fixed

hategan commented:
For the record, while updating the CSS, I ran into an issue in which the branch indicator divs were showing even when they were disabled. This occurred after switching views out and back into the calendar view.

aschwanden1 commented:
> For the record, while updating the CSS, I ran into an issue in which the branch indicator divs were showing even when they were disabled. This occurred after switching views out and back into the calendar view.

Addressed and fixed

hategan commented on src/psij/web/scss/style.scss:
```diff
@@ -1,15 +1,72 @@
+    text-align: right;
+    margin-bottom: 32pt;
+    background-color: #FEB88B;
+    background-image: linear-gradient(180deg, #FEB88B 0%, $exaRed 74%);
```
The styling was reverted to the current styling in main

hategan commented on src/psij/web/js/SitesModel.js:
```diff
@@ -0,0 +1,381 @@
+    var getModel_ = function( callback ) {
+
+        callback_ = callback;
+        getLatestResultForTestOnAllSites_();
```
It seems that the curated sites view uses both: the existing endpoint to show the status of the test suite and the tests endpoint to show the status of individual tests.

hategan commented on src/psij/web/js/BuildAccordianModelFromSites.js:
```diff
@@ -0,0 +1,96 @@
+ */
+PS.BuildAccordianModelFromSites = function() {
+
+    var get_states_ = function( months ) {
```
Well, we need to maintain this code, and having consistent styling and naming conventions help with that so that, in the long run, there is less effort in maintaining the product so that effort can be focused on other things, such as features.

aschwanden1 replied:
Agreed.  from now on, I'll try to remember no spaces after or before a ( ).

hategan replied:
Should fix in the future: #42.

hategan commented on src/psij/web/js/BuildAccordianModelFromSites.js:
```diff
@@ -0,0 +1,96 @@
+    };
+
+
+    var get_ = function( accordianModelRef ) {
```
Have you read this? https://v2.vuejs.org/v2/guide/reactivity.html

aschwanden1 replied:
I'll take a look at that.  Thanks for the link.

hategan commented on src/psij/web/js/BuildAccordianModelFromSites.js:
```diff
@@ -0,0 +1,96 @@
+
+            machObjs[domain] = machObjs[domain] || [];
+            machObjs[domain].push({
+                machine: site.site_id,
```
I don't think anything has changed:
https://github.com/ExaWorks/psij-testing-service/blob/pa-accordian/src/psij/web/js/BuildAccordionModelFromSites.js#L60

aschwanden1 replied:
ok, now it should be ok.

hategan replied:
Sorry for not being clear, but `domain` was fine.  `siteId` is clearly not the site id.
'machine' on the other hand, it is not quite clear why we need a new term for what is the site id, so that should be changed.

I didn't look at the snippet above. I checked the actual file in the branch and the change wasn't there.

hategan replied:
You have:
```
var siteId = globalMethods.getDomain(site.site_id);
```

the variable `siteId` is not the same as `site.site_id`. You call `getDomain` to get the domain from the site id, so whatever is returned should not be put in something called `siteId` because it is NOT the same as the site id.

hategan commented on src/psij/web/js/BuildAccordianModelFromSites.js:
```diff
@@ -0,0 +1,96 @@
+        for( var domain in machObjs ) {
+
+            am.push({
+                logo_name: domain,
```
Same as above. Does not seem to have changed.

aschwanden1 replied:
should be good now.

hategan replied:
`logo_name` should be `domain`.
I'm not quite sure what `logo_name` is supposed to mean.

aschwanden1 replied:
logo_name is the thing we're showing in the context of the view.  sometimes it's not necessarily a domain that we're showing but for example "Hategan's laptop".  therefore, skipping.

hategan replied:
Sounds like `display_name` is what might fit better. Anyway, not important.

hategan commented on src/psij/web/js/SitesModel.js:
```diff
@@ -0,0 +1,381 @@
+
+            var machinesInAccordian = [];
+
+            CS.yaxis[yIndex].horizontalSummary = makeGoodStatesArray_(CS.xaxis.length);
```
I think that's the point of the comment. There is no "default value" when we're talking about test results. It should only be the test results. Showing "passed" when something goes wrong is a really really bad idea.

aschwanden1 replied:
No, if something goes wrong, then it shows up red.

This is the paradigm we agreed to in the meeting: By default for summary values we show green unless a test fails then we show red.

hategan commented on src/psij/web/js/SitesModel.js:
```diff
@@ -0,0 +1,381 @@
+                for( var y=0; y < states.length; y++ ) {
+                    objStates.push({
+                       mainState: states[y],
+                       branches: []
```
Ok, can you point to where that field is populated? I searched for all occurrences of ".branches", but none of them were assignments.

aschwanden1 replied:
branches gets used on RowAccordion.js Line 40 for the mini branches.

the data gets populated at BuildAccordionModelFromsites.js Line 38.

hategan replied:
In BuildAccordionModelFromsites.js get_states (which contains Line 38) it constructs a completely new objStates array, so I do not see how the empty array above is modified for it to count as a placeholder.
So maybe this is used to show empty bubbles before data is loaded. Moving on...

hategan commented on src/psij/web/summary.html:
```diff
@@ -73,193 +92,181 @@
+
+                <div v-for="yaxe in accordianModel">
+                    <row-accordian :yaxe_prop="yaxe"></row-accordian>
+                </div>
```
I'm not sure how your reply is related to my initial comment.

Would you like me to file issues for the those bullet points?

aschwanden1 replied:
Yes, Please.

hategan commented on src/psij/web/scss/RowAccordian.scss:
```diff
@@ -0,0 +1,107 @@
+.RowAccordian {
```
I don't think consistency with a file that was added as part of this PR take precedence over existing "naming conventions or at least follow generally accepted standards."

hategan commented on src/psij/web/scss/style.scss:
```diff
@@ -142,75 +195,64 @@ body {
 }
 
-
+#status-table .site-branch-cell {
```
I'm assuming you mean that style.scss was removed.
